### PR TITLE
OJO data sample

### LINKED ISSUE #16: Generate sample of ojo job adverts

### PR DESCRIPTION
**Pull request by lizgzil**
---

# Description

Find which job adverts are duplicates, then create a deduplicated sample of the data.

This has been outputted here `s3://prinz-green-jobs/outputs/data/ojo_application/deduplicated_sample/ojo_sample.csv` .

Fixes #16 

# Instructions for Reviewer

To first find which job adverts in the OJO database are duplicates run:

```
dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication.py
```

this creates a csv with deduplicated job adverts ids in a certain time window. The default is a 7 day window; so this means if two adverts with the same text and location turn up within a week of each other, then only one will remain in the outputted data.

## Sampling

To sample the deduplicated dataset run:

```
python dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/sample_ojo.py
```

this will create a sample of the deduplicated OJO dataset and output all the OJO tables filtered by this sample. All outputs will go to the `s3://prinz-green-jobs/outputs/data/ojo_application/deduplicated_sample/` S3 folder. The main file of interest is `s3://prinz-green-jobs/outputs/data/ojo_application/deduplicated_sample/ojo_sample.csv` which contains the job advert descriptions, the location, date, and the job title.

# Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [ ] I have checked the code runs
- [ ] I have tested the code
- [ ] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [ ] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [ ] I have explained this PR above
- [ ] I have requested a code review

### CODE REVIEW

**Comment by lizgzil** [comment]
In this sample there are 94955 unique descriptions out of 100,000. This will be because the description is the same, but they weren't published in the same time window OR the location is different

File: dap_prinz_green_jobs/getters/ojo.py [hunk 1]
```diff
@@ -15,4 +15,8 @@ def get_ojo_sample() -> pd.DataFrame:
     Returns:
         pd.Dataframe: ojo sample data
     """
-    return load_s3_data(s3, BUCKET_NAME, "outputs/data/ojo_application/ojo_sample.csv")
```
Comment (india-kerle) [inline dap_prinz_green_jobs/getters/ojo.py:18]: we could have getters for the other outputs i.e. 

def get_ojo_sample_skills() -> pd.DataFrame: .....

File: dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/README.md [hunk 2]
```diff
@@ -1,3 +1,25 @@
...
 
-This directory will contain scripts to generate samples of OJO data.
+This directory contains scripts to generate samples of OJO data.
+
+## Deduplication
+
+To first find which job adverts in the OJO database are duplicates run:
+
+```
+dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication.py
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/README.md]:
```suggestion
python dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication.py
```

Comment (india-kerle) [reply dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/README.md]: this ran for me!

File: dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/README.md [hunk 3]
```diff
@@ -1,3 +1,25 @@
...
+```
+
+this creates a csv with deduplicated job adverts ids in a certain time window. The default is a 7 day window; so this means if two adverts with the same text and location turn up within a week of each other, then only one will remain in the outputted data.
+
+## Sampling
+
+To sample the deduplicated dataset run:
+
+```
+python dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/sample_ojo.py
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/README.md:21]: this also ran for me 👍

File: dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+from tqdm import tqdm
+
+from dap_prinz_green_jobs.getters.data_getters import save_to_s3, get_s3_resource
+from dap_prinz_green_jobs import BUCKET_NAME, logger, config
+from dap_prinz_green_jobs.pipeline.ojo_application.ojo_sample.deduplication_utils import (
+    short_hash,
+    get_deduplicated_job_adverts,
+)
+
+s3 = get_s3_resource()
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication.py]: part of me thinks we put this in data_getters.py so we aren't constantly re-creating this object all over scripts in the codebase

File: dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
+"""
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication.py:1]: looks great! 💯

File: dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication_utils.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+from datetime import datetime, timedelta, date
+from dateutil.relativedelta import relativedelta
+import random
+from hashlib import md5
+
+from dap_prinz_green_jobs import logger
+
+
+def short_hash(text):
+    """Generate a unique short hash for this string - from ojd_daps"""
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication_utils.py]: being pedantic but you could add types i.e.  
```suggestion
def short_hash(text:str) -> int:
    """Generate a unique short hash for this string - from ojd_daps"""
```

File: dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication_utils.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    hx_code = md5(text.encode()).hexdigest()
+    int_code = int(hx_code, 16)
+    short_code = str(int_code)[:16]
+    return int(short_code)
+
+
+def get_date_ranges(start, end, num_units=7, unit_type="days"):
```
Comment (india-kerle) [inline dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication_utils.py]: ditto types for the rest of the functions - but your doc strings have that context so up to you!

**Review by india-kerle — COMMENTED** [review]
looks great! I was able to run both scripts. My only observation would be to add additional getters in ojo.py based on the sample so we can just load the other types of metadata using getters.

File: dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication.py (hunk 4 above)
Comment (lizgzil) [reply dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication.py]: makes a lot of sense!

File: dap_prinz_green_jobs/getters/ojo.py (hunk 1 above)
Comment (lizgzil) [reply dap_prinz_green_jobs/getters/ojo.py:18]: I've added these to ojo.py now 👍

File: dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication.py (hunk 4 above)
Comment (lizgzil) [reply dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication.py]: I've done it! I've made the relevant changes across the repo here [f06b807](https://github.com/nestauk/dap_prinz_green_jobs/pull/19/commits/f06b8076eb211f923a3af71148d28e7960cc5635) just need to remember to not add the s3 argument to the load and save functions now!

File: dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication_utils.py (hunk 6 above)
Comment (lizgzil) [reply dap_prinz_green_jobs/pipeline/ojo_application/ojo_sample/deduplication_utils.py]: I am so bad for this! but really want to start with good practices!

**Comment by lizgzil** [comment]
thanks @india-kerle !

**Pull request merged**